### PR TITLE
fix(storage): catch http timeout error and retry 

### PR DESCRIPTION
**Pull request by wcy-fdu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by Little-Wallace — COMMENTED** [review]
Do we need to support retry for `put_object` ?

**Comment by wcy-fdu** [comment]
> Do we need to support retry for `put_object` ?

This is a workaround for the `get_object` timeout error, so we only catch the special `DispatchFailure::ConnectError`, for write, we can still use `RetryConfig` provided by aws itself. Note that this PR may not solve the problem, so wait a while after it is merged to see if the error still occurs.

File: src/object_store/src/object/s3.rs [hunk 1]
```diff
@@ -702,6 +726,24 @@ impl S3ObjectStore {
             }
         }
     }
+
+    #[inline(always)]
+    fn get_retry_strategy() -> Map<Take<ExponentialBackoff>, fn(Duration) -> Duration> {
+        ExponentialBackoff::from_millis(DEFAULT_RETRY_INTERVAL)
```
Comment (Little-Wallace) [inline src/object_store/src/object/s3.rs:732]: Does it means that any request which exceed 20ms would send another request ?
20ms is too small for s3 storage. 
When we read a large object ,the latency could be several hundred mills

Comment (wcy-fdu) [reply src/object_store/src/object/s3.rs:732]: No, only timeout error occurs will will retry, see `should_retry`.

**Comment by wcy-fdu** [comment]
@hzxa21  PTAL❤️

File: src/object_store/src/object/s3.rs [hunk 2]
```diff
@@ -702,6 +726,24 @@ impl S3ObjectStore {
...
+            .max_delay(DEFAULT_RETRY_MAX_DELAY)
+            .take(DEFAULT_RETRY_MAX_ATTEMPTS)
+            .map(jitter)
```
Comment (hzxa21) [inline src/object_store/src/object/s3.rs:735]: FYI, after checking the implementation of `ExponentialBackoff`, with DEFAULT_RETRY_INTERVAL=20ms, DEFAULT_RETRY_MAX_DELAY=5s, DEFAULT_RETRY_MAX_ATTEMPTS=10, I think the current behavior is to generate the following backoff time for each attempt 
```
[20ms * rand(), 400ms * rand(), 5s * rand(), 5s * rand(), 5s * rand(), 5s * rand(), 5s * rand(), 5s * rand(), 5s * rand(), 5s * rand()]
```

I believe when timeout happens, we will most likely hit the `5s * rand()` under the current configs. I am uncertain about whether this is appropriate but it is okay to give it a try.

**Review by hzxa21 — APPROVED** [review]
We can merge the PR and monitor `object_store_failure_count` metrics to verify whether this PR works.

File: src/object_store/src/object/s3.rs (hunk 2 above)
Comment (wcy-fdu) [reply src/object_store/src/object/s3.rs:735]: Or maybe we can change `DEFAULT_RETRY_MAX_DELAY = 10s` to make the backoff time more values?
`[20ms * rand(), 400ms * rand(), 8s * rand(), 10s * rand(), 10s * rand(), 10s * rand(), 10s * rand(), 10s * rand(), 10s * rand(), 10s * rand()]`
I think whether peak clipping is possible depends on the coefficient of rand.

**Pull request merged**